refactor(major): Refactor entire india banking client side flows

Bhavan23 opened the pull request.

Bhavan23 committed:
refactor: enable payment request creation for adhoc payment.

Bhavan23 committed:
refactor: override the payment request's "make_payment_request_function"

Bhavan23 committed:
refactor: remove the overrite payment request.

Bhavan23 committed:
refactor: Create a custom payment field for both payment orders and requests.

Bhavan23 committed:
refactor: enable create payment request button in purchase order and purchase invoice

Bhavan23 committed:
refactor: modify payment_order_type option with journal entries

Bhavan23 committed:
refactor: add default bank

Bhavan23 committed:
refactor: create default mode of transfer

Bhavan23 committed:
refactor: add default payment type

Bhavan23 committed:
refactor: Place uninstall hooks in a different file.

Bhavan23 committed:
fix: remove payment order copy file
refactor: update install file path

Bhavan23 committed:
refactor: take out the bank event and add it to the bank overrite class.

Bhavan23 committed:
fix: remove overrides bank and use docevent hooks

Bhavan23 committed:
refactor: payments button label and api path

Bhavan23 committed:
refactor: remove purchase order and purchase invoice js overrides

Bhavan23 committed:
fix: create default mode of transfers

Bhavan23 committed:
fix: remove patch file

Bhavan23 committed:
refactor: remove unused function and imports from utils

Bhavan23 committed:
refactor: remove unused api file

Bhavan23 committed:
refactor: format indentation

Bhavan23 committed:
refactor: remove unused file from doc_events

Bhavan23 committed:
refactor: create custom fields

Bhavan23 committed:
refactor: remove customisation exports

Bhavan23 committed:
refactor: remove bank payment request doc

Bhavan23 committed:
safty commit

Bhavan23 committed:
refactor: bulk payment code add in bank connector
refactor: single payment code add in bank connector

Bhavan23 committed:
refactor: revamped payment status code in bank connector

Bhavan23 committed:
fix: convert the request log to look prettier
fix: add filter to the bank account
fix: fetch company bank account details during the API call

Bhavan23 committed:
fix: handle payment response json validation erro
fix: rename method action in payload

Bhavan23 committed:
refactor: refactor and optimize

Bhavan23 committed:
typo: remove debugger-throw on payment order validation

Bhavan23 committed:
feat: implement workflow enabling/disabling functionality in IndiaBankingSettings

Bhavan23 committed:
safety_commit

Bhavan23 committed:
fix: change variable name to python snake case
fix: add diabled filter in default mode of payment field

Bhavan23 committed:
Merge pull request #12 from aerele/safety-commit-new

safety_commit

Bhavan23 committed:
refactor: add summarise based on field
fix: remove is party wise field

Bhavan23 committed:
refactor:  modify all is party-wise condition to summarise based on condition

Bhavan23 committed:
refactor: pretify

Bhavan23 committed:
feat: update dimension while summarising the payment order

Bhavan23 committed:
refactor: add default bankaccount validation
fix: add company bank account filter

Bhavan23 committed:
refactor: modify SQL query using query builder

Bhavan23 committed:
fix: use bank account filter to fetch bank entry
refactor: update accounting dimention

Bhavan23 committed:
refactor: recreate Bank Payment Request with disabled create options

Bhavan23 committed:
refactor: prettify the payment request js code

Bhavan23 committed:
refactor: remove the bank account fixture workflow
refactor: add script to create the bank account workflow

Bhavan23 committed:
refactor: optimise bank doc_events code

Bhavan23 committed:
fix: remove payment order from accounting dimention hook

Bhavan23 committed:
fix: update payment status before reinitiate payment.

Bhavan23 committed:
refactor: while create api log return name ref
refactor: add translate

Bhavan23 committed:
fix: add summarize based on only draft state
refactor: pretify message content

Bhavan23 committed:
feat: add pending payment cancel button
fix: add missing fields

Bhavan23 committed:
refactor: add custom field update patch

Bhavan23 committed:
fix: while create payment entry modify filter condition with accounting dimention

Bhavan23 committed:
feat: set party field name for share holder and customer party type

Bhavan23 committed:
fix: reload doc once value update

Bhavan23 committed:
refactor: optimise 'validate_summary' method

Bhavan23 committed:
refactor: optimise update payment status method
refactor: removed unused function 'get_bank_entry_for_payroll'

Bhavan23 committed:
refactor: optimise payment entry py file

Bhavan23 committed:
refactor: optimise payment request python code

Bhavan23 committed:
refactor: add translate to all message
refactor: optimise payment request overrides code

Bhavan23 committed:
fix: while process payment request summarise field extend with payment entry and journal entry account

Bhavan23 committed:
refactor: move "get_bank_balance" function in bank connector

Bhavan23 committed:
refactor: remove unused function from payment_order python file

Bhavan23 committed:
refactor: stop bank payment request client actions

Bhavan23 committed:
refactor: stop bank payment request client actions

Bhavan23 committed:
typo: use js multiline string syntax instead of useing "`"

Bhavan23 committed:
fix: add payment create_payment_entry function in payment order

Bhavan23 committed:
fix: update Approved status as completed

Bhavan23 committed:
fix: remove doc.reload avoid data loss
fix: add condition payment order type

Bhavan23 committed:
feat: add payment in the background useing enable_payment_in_the_background

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,105 @@
+		"maximum_limit": 200000,
+		"start_time": "0:00:00",
+		"end_time": "23:59:59",
+		"disabled": 1,
```
Why is every mode of transfer, by default disabled?

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,105 @@
+	{
+		"mode": "RTGS",
+		"minimum_limit": 200000,
+		"maximum_limit": 50000000,
```
Make 50 Crore as the upper limit, beyond that LEI number is required.

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,105 @@
+	{
+		"mode": "NEFT",
+		"minimum_limit": 0,
+		"maximum_limit": 100000000000,
```
Make 50 Crore as the upper limit, beyond that LEI number is required.

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,105 @@
+		"start_time": "0:00:00",
+		"end_time": "23:59:59",
+		"disabled": 1,
+		"priority": "1",
```
This is account specific. Please bring in that key here. And do the following priority.

1) A2A
2) IMPS
3) RTGS
4) NEFT

by default. Let the user re-order as per their need.

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,105 @@
+]
+
+STD_BANK_LIST = [
+	"Yes Bank",
```
Remove Yes Bank as of now. Reorder this in alphabetical order.

Vigneshsekar commented on india_banking/hooks.py:
```diff
@@ -4,6 +4,46 @@
+
+after_install = ["india_banking.install.after_install"]
+
+before_uninstall = "india_banking.uninstall.before_uninstall"
```
Can you put this in array too? Is it supported?

Bhavan23 replied:
yes, but i modify because it is only one path.

Vigneshsekar commented on india_banking/hooks.py:
```diff
@@ -4,6 +4,46 @@
+doc_events = {
+	"Bank": {"on_trash": "india_banking.india_banking.doc_events.bank.bank_on_trash"},
+	"Bank Account": {
+		"validate": "india_banking.india_banking.doc_events.bank.validate_ifsc_code"
```
Make a new file for bank_account, as it is a separate doctype.

Vigneshsekar commented on india_banking/hooks.py:
```diff
@@ -4,6 +4,46 @@
+}
+
+doc_events = {
+	"Bank": {"on_trash": "india_banking.india_banking.doc_events.bank.bank_on_trash"},
```
Can we rename this function as disallow_standard_bank_deletion

Vigneshsekar commented on india_banking/india_banking/doctype/bank_connector/bank_connector.js:
```diff
@@ -8,7 +8,8 @@ frappe.ui.form.on("Bank Connector", {
 					is_default: 1,
-                    is_company_account: 1
+                    is_company_account: 1,
+					company: doc.company,
```
Please use the standard linter and rectify the spacing for the entire repository.

Vigneshsekar commented on india_banking/india_banking/doc_events/bank.py:
```diff
@@ -0,0 +1,19 @@
+
+
+def validate_ifsc_code(doc, method=None):
+	if not IFSC_PATTERN.match(cstr(doc.branch_code)):
```
User often get confused with branch code and IFSC code. Can we set the label as `IFSC Code` for this field if this app is installed?

Bhavan23 committed:
style: format codebase with linter

Bhavan23 committed:
refactor: enable mode of transfer while installing
chore: rename 'bank_on_trash' to 'disallow_standard_bank_deletion'
fix: update maximum limit for all mode of transfer
chore: rename 'Branch Code' label to 'IFSC Code' in bank account doc

Bhavan23 committed:
refactor: add translate to message

Bhavan23 committed:
fix: include payload with summary row details for payment reference

Bhavan23 committed:
feat: In the log doc, add a button to display the failure reason.

Vigneshsekar commented on india_banking/default.py:
```diff
@@ -0,0 +1,103 @@
+]
+
+STD_BANK_LIST = [
+	"HDFC Bank",
```
Why is Axis Bank, removed?

Vigneshsekar commented on india_banking/install.py:
```diff
@@ -0,0 +1,603 @@
+				"insert_after": "posting_date",
+			},
+			{
+				"label": "ICICI Bank Api Info",
```
Bank specific fields can be avoided. What shall we do here?

Vigneshsekar commented on india_banking/install.py:
```diff
@@ -0,0 +1,603 @@
+			"fieldname": "grand_total",
+			"property": "reqd",
+			"property_type": "Check",
+			"value": 0,
```
Why is it set to 0?, Is the grand_total not mandatory?

Bhavan23 replied:
The grand total will be calculated using the net total. we keep net total as mandatory

Vigneshsekar commented on india_banking/install.py:
```diff
@@ -0,0 +1,603 @@
+
+def create_default_payment_type():
+	if not frappe.db.exists("Payment Type", "Pay"):
+		frappe.get_doc({"doctype": "Payment Type", "payment_type": "Pay"}).insert(
```
Instead of ignoring mandatory, can we not get the default accounts payable account from the company and set it to the payment type?

Vigneshsekar commented on india_banking/public/js/payment_request.js:
```diff
@@ -1,71 +1,66 @@
+
+function get_bank_query_conditions(frm) {
+  let conditions = {
+    company: frm.doc.company,
```
Would you also like to consider the workflow state as `approved` here?
And is company filter required? Does bank account have company and is it mandatory?

Vigneshsekar commented on india_banking/public/js/payment_order.js:
```diff
@@ -1,363 +1,512 @@
+      get_query_filters: {
+        docstatus: 1,
+        status: ["=", "Initiated"],
+        bank: frm.doc.bank,
```
What is this condition?

Vigneshsekar replied:
bank: frm.doc.bank,

Vigneshsekar commented on india_banking/public/js/payment_order.js:
```diff
@@ -1,363 +1,512 @@
+          query: "india_banking.overrides.journal_entry.get_bank_entry",
+          filters: {
+            docs: existing_journal_entries,
+            company_account: frm.doc.account,
```
posting_date lesser than current date, currency being INR, only those lines that are involving the company bank account of the payment order, are these handled?

Vigneshsekar commented on india_banking/public/js/payment_order.js:
```diff
@@ -1,363 +1,512 @@
+        source_doctype: ["!=", "Payment Request"],
+        payment_type: "Pay",
+        mode_of_payment: "Wire Transfer",
+        bank_account: frm.doc.company_bank_account,
```
posting_date lesser than current date, currency being INR, are these conditions taken care of?

Vigneshsekar commented on india_banking/public/js/bank_account.js:
```diff
@@ -1,32 +1,37 @@
+  refresh(frm) {
+    if (
+      frm.doc.is_company_account &&
+      frm.doc.is_company_account &&
```
Checked twice

Vigneshsekar commented on india_banking/patches/v1_0/update_custom_fields.py:
```diff
@@ -0,0 +1,42 @@
+		},
+		pluck="name",
+	):
+		frappe.db.delete("Custom Field", {"name": ["in", fields]})
```
Does it not delete the ones that are not created by the India banking application too? This seems dangerous.

Bhavan23 replied:
reverted

Vigneshsekar commented on india_banking/patches/v1_0/update_custom_fields.py:
```diff
@@ -0,0 +1,42 @@
+		"Payment Request",
+	]
+	for doctype in doctypes:
+		frappe.db.delete("Property Setter", {"name": f"{doctype}-main-field_order"})
```
This too, this is also not verifiying any. Seems dangerous.

Bhavan23 replied:
reverted

Vigneshsekar commented on india_banking/overrides/payment_order.py:
```diff
@@ -0,0 +1,268 @@
+						"You cannot cancel a payment order with Initiated/Processed payments"
+					)
+				)
+		super().on_cancel()
```
Is this required?

Bhavan23 replied:
on_cancel is already there in the ERPNext app. They are using on_cancel to update the status.

Vigneshsekar commented on india_banking/overrides/payment_entry.py:
```diff
@@ -0,0 +1,89 @@
+from frappe import _
+
+
+class CustomPaymentEntry(PaymentEntry):
```
As you're not overriding anything here, then this class and file is not required? Can you move the below code elsewhere and remove this class and file.

Vigneshsekar commented on india_banking/overrides/journal_entry.py:
```diff
@@ -0,0 +1,147 @@
+			)
+			.where(
+				(JournalEntryAccount.parent == source.name)
+				& (JournalEntryAccount.party_type == "Employee")
```
Does it mean, we do not allow any supplier, customer bank entry type journal entries?

Vigneshsekar commented on india_banking/overrides/journal_entry.py:
```diff
@@ -0,0 +1,147 @@
+			"account",
+			"cost_center",
+			"project",
+			"debit as amount",
```
Assuming debit and debit in account currency as same here? Can you add currency as INR as a validation in the query?

Vigneshsekar commented on india_banking/india_banking/doctype/india_banking_request_log/india_banking_request_log.py:
```diff
@@ -1,38 +1,84 @@
 	else:
-		frappe.db.commit()
\ No newline at end of file
+		frappe.db.commit()
```
Why are you manually commiting?

Bhavan23 replied:
To avoid request and response data loss while processing the response, some time will raise an error.

Bhavan23 committed:
feat: create default payment type for company
refactor: customisation added

Bhavan23 committed:
refactor: remove custom field deletetion without checking filter.

Bhavan23 committed:
chore: rename bank_balance key to get_bank_balance

Bhavan23 committed:
fix: change 'or' condition instead of 'and'

Bhavan23 committed:
fix: update currency filter for payment cycle

Bhavan23 committed:
refactor: remove unique id and filereference id fields

Bhavan23 committed:
refactor: modify the payment api structure

Bhavan23 committed:
feat: add settings to reposting entry

Bhavan23 committed:
feat: handle unreconcil payment
